pkg/controller/controller.go

```diff
@@ -0,0 +1,149 @@
+package controller
+
+import (
+	kcache "github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	kutil "github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// RunnableController is a controller which implements a Run loop.
+type RunnableController interface {
+	// Run starts the asynchronous controller loop.
+	Run()
+}
+
+// RetryController is a RunnableController which delegates resource
+// handling to a function and knows how to safely manage retries of a resource
+// which failed to be successfully handled.
+type RetryController struct {
+	// Queue is where work is retrieved for Handle.
+	Queue Queue
+
+	// Handle is expected to process the next resource from the queue.
+	Handle func(interface{}) error
+
+	// ShouldRetry returns true if the resource and error returned from
+	// HandleNext should trigger a retry via the RetryManager.
+	ShouldRetry func(interface{}, error) bool
+
+	// RetryManager is fed the handled resource if Handle returns a Retryable
+	// error. If Handle returns no error, the RetryManager is asked to forget
+	// the resource.
+	RetryManager RetryManager
+}
+
+// Queue is a narrow abstraction of a cache.FIFO.
+type Queue interface {
+	Pop() interface{}
+	AddIfNotPresent(interface{}) error
+}
+
+// RetryOnRetryableError is a simple ShouldRetry implementation which returns
+// true if the error is a RetryableError.
+var RetryOnRetryableError = func(resource interface{}, err error) bool {
+	if _, isRetryable := err.(*RetryableError); isRetryable {
+		return true
+	}
+	return false
+}
+
+// Run begins processing resources from Queue asynchronously.
+func (c *RetryController) Run() {
+	go kutil.Forever(func() { c.handleOne(c.Queue.Pop()) }, 0)
+}
+
+// handleOne processes resource with Handle. If Handle returns a retryable
+// error, the handled resource is passed to the RetryManager. If no error is
+// returned from Handle, the RetryManager is asked to forget the processed
+// resource.
+func (c *RetryController) handleOne(resource interface{}) {
+	err := c.Handle(resource)
+	if err != nil {
+		if c.ShouldRetry(resource, err) {
+			c.RetryManager.Retry(resource)
+			return
+		}
+	}
+	c.RetryManager.Forget(resource)
+}
+
+// RetryableError is an error that is recoverable and which should
+// signal another attempt at processing a resource.
+type RetryableError struct {
+	err error
+}
+
+func (e *RetryableError) Error() string {
```

Comment:
If someone using the RetryController wants to establish their own error taxonomy and retry handler around it, they're free to do so and the abstractions here will facilitate them. However, all our existing use cases in deployment and builds (I think) can use this OOTB error type and handler for now, IMO.

I could me misunderstanding the proposed alternative design... maybe try once more to clarify just to be sure?